lib/engine-jit/src/serialize.rs

```diff
@@ -38,9 +32,76 @@ pub struct SerializableCompilation {
 
 /// Serializable struct that is able to serialize from and to
 /// a `JITArtifactInfo`.
-#[derive(Serialize, Deserialize, MemoryUsage)]
+#[derive(Serialize, Deserialize, MemoryUsage, Archive, RkyvDeserialize, RkyvSerialize)]
 pub struct SerializableModule {
     pub compilation: SerializableCompilation,
     pub compile_info: CompileModuleInfo,
     pub data_initializers: Box<[OwnedDataInitializer]>,
 }
+
+fn to_serialize_error(err: impl std::error::Error) -> SerializeError {
+    SerializeError::Generic(format!("{}", err))
+}
+
+impl SerializableModule {
+    /// Serialize a Module into bytes
+    /// The bytes will have the following format:
+    /// RKYV serialization (any length) + POS (8 bytes) + Endian (1 byte)
+    pub fn serialize(&self) -> Result<Vec<u8>, SerializeError> {
+        let mut serializer = SharedSerializerAdapter::new(WriteSerializer::new(vec![]));
+        let pos = serializer
+            .serialize_value(self)
+            .map_err(to_serialize_error)? as u64;
+        let mut serialized_data = serializer.into_inner().into_inner();
+        serialized_data.extend_from_slice(&pos.to_le_bytes());
+        if cfg!(target_endian = "big") {
+            serialized_data.extend_from_slice(&[b'b']);
+        } else if cfg!(target_endian = "little") {
+            serialized_data.extend_from_slice(&[b'l']);
+        }
+        Ok(serialized_data)
+    }
+
+    /// Deserialize a Module from a slice.
+    /// The slice must have the following format:
+    /// RKYV serialization (any length) + POS (8 bytes) + Endian (1 byte)
+    ///
+    /// # Safety
+    /// This method is highly unsafe since we are deserializing directly
+    /// from memory, without validating it first.
+    pub unsafe fn deserialize(metadata_slice: &[u8]) -> Result<Self, DeserializeError> {
+        let archived = Self::archive_from_slice(metadata_slice)?;
+        Self::deserialize_from_archive(archived)
+    }
+
+    unsafe fn archive_from_slice<'a>(
+        metadata_slice: &'a [u8],
+    ) -> Result<&'a ArchivedSerializableModule, DeserializeError> {
+        if metadata_slice.len() < 9 {
+            return Err(DeserializeError::Incompatible(
+                "invalid serialized data".into(),
+            ));
+        }
+        let mut pos: [u8; 8] = Default::default();
+        let endian = metadata_slice[metadata_slice.len() - 1];
+        if (cfg!(target_endian = "big") && endian == b'l')
+            || (cfg!(target_endian = "little") && endian == b'b')
```

Comment:
Be aware that this logic may not do exactly what you want in degenerate cases.

Having the byte `b'j'` and `target_endian = "big"` is a success case here. (i.e. it only checks for well-defined failure: assuming only 2 possible bytes without validating it)

Reply:
Fixed!